### PR TITLE
Change handling of theme defaults for glyphs by using mixin classes (#9693 and #9339)

### PR DESCRIPTION
**Pull request by jeremy9959**
This responds to the discussion in issue #9693 and #9339 with @bryevdv and others. The proposal is to create four new abstract classes: 
- LineGlyph
- FillGlyph
- TextGlyph 
- HatchGlyph

Glyphs that support the line properties would inherit LineGlyph; those that support text properties would inherit from TextGlyph; and so on.  Then a theme like this:
```
theme_json = {
    "attrs": {
        "LineGlyph": {"line_color": "#ee33ee", "line_width": 2},
        "FillGlyph": {"fill_color": "orange",},
        "HatchGlyph": {"hatch_pattern":"@","hatch_alpha": .8,},
        "TextGlyph": {
            "text_color": "red",
            "text_font_style": "bold",
            "text_font": "Helvetica",
        },
        "Ellipse": {"fill_color": "green", "line_color": "yellow","fill_alpha":.2},
    }
}
```
would set colors, fills, and hatch patterns for glyphs that support them; the explicit options for Ellipse would override these.

Here is some demonstration code:
```
from bokeh.models import (
    ColumnDataSource,
    Title,
    Ellipse,
    HBar,
    Text,
    Plot,
    LinearAxis,
    DataRange1d,
    Scatter,
    Line,
)
from bokeh.models.tickers import BasicTicker
from bokeh.models.formatters import BasicTickFormatter
from bokeh.transform import dodge
from bokeh.themes import Theme
from bokeh.io import output_file, show, curdoc
import numpy as np

output_file("./minimal.html")

theme_json = {
    "attrs": {
        "Plot": {"width": 400, "height": 400, "background_fill_color": "#eeeeee"},
        "Grid": {"visible": False},
        "Title": {"text": "Demo of Themes"},
        "LinearAxis": {
            "axis_line_color": "#ffffff",
            "major_tick_line_color": "#ffffff",
            "axis_label_text_font_size": "10pt",
            "axis_label_text_font_style": "bold",
        },
        "LineGlyph": {"line_color": "#ee33ee", "line_width": 2},
        "FillGlyph": {"fill_color": "orange"},
        "HatchGlyph": {"hatch_pattern": "@", "hatch_alpha": 0.8},
        "TextGlyph": {
            "text_color": "red",
            "text_font_style": "bold",
            "text_font": "Helvetica",
        },
        "Ellipse": {"fill_color": "green", "line_color": "yellow", "fill_alpha": 0.2},
    }
}

source = ColumnDataSource(
    {"x": [1, 2, 3, 4, 5], "y": [1, 2, 3, 4, 5], "who": ["a", "b", "c", "d", "e"]}
)
x = np.linspace(1, 5, 100)
y = x + np.sin((x - 1) * np.pi)
x2 = np.linspace(1.5, 5.5, 5)
z = x2 + 2 * np.cos((x2 - 1) * np.pi)
source2 = ColumnDataSource({"x": x, "y": y})
source3 = ColumnDataSource({"x": x2, "y": z})


plot = Plot(
    plot_width=300, plot_height=300, x_range=DataRange1d(), y_range=DataRange1d()
)
plot.title = Title()
xaxis = LinearAxis(ticker=BasicTicker(), formatter=BasicTickFormatter())
yaxis = LinearAxis(ticker=BasicTicker(), formatter=BasicTickFormatter())
plot.add_layout(xaxis, "below")
plot.add_layout(yaxis, "left")

plot.add_glyph(source, Scatter(x="x", y="y", marker="diamond", size=20))
plot.add_glyph(source3, Ellipse(x="x", y="y", width=0.2, height=0.3, angle=-0.7))
plot.add_glyph(source, Text(x=dodge("x", -0.2), y=dodge("y", 0.1), text="who"))
plot.add_glyph(source2, Line(x="x", y="y"))
plot.add_glyph(
    ColumnDataSource({"y": [2.5], "x": [0.5]}), glyph=HBar(y="y", right="x", height=1.5)
)
curdoc().theme = Theme(json=theme_json)
show(plot)
```
which yields this output:
Screen Shot 2020-03-02 at 4 29 19 PM

One concern: this appears to slow things down, possibly because traversing the more complicated class hierarchy to set defaults hits performance.  But this is an unscientific impression.

### CODE REVIEW

**Comment by jeremy9959** [comment]
Unit tests are failing because of changes to glyph classes.

File: bokeh/models/glyphs.py [hunk 1]
```diff
@@ -737,580 +1027,835 @@ class MultiPolygons(Glyph):
...
 
-    fill_props = Include(FillProps, use_prefix=False, help="""
+    fill_props = Include(
+        FillProps,
+        use_prefix=False,
+        help="""
     The %s values for the wedges.
-    """)
+    """,
+    )
```
Comment (mattpap) [inline bokeh/models/glyphs.py]: Why all those unrelated changes? How we are supposed to know what you actually changed in this file? You will have to revert unrelated changes and leave the minimum number of changes to resolve the issues in question.

Comment (jeremy9959) [reply bokeh/models/glyphs.py]: I applied black formatting unconsciously, I hopefully cleaned it up and it is clearer now.

Comment (mattpap) [reply bokeh/models/glyphs.py]: Thanks. It's much clearer now to see what's the extent of your changes.

File: bokeh/models/glyph.py [hunk 2]
```diff
@@ -120,6 +120,26 @@ class ConnectedXYGlyph(XYGlyph):
 
     '''
 
+@abstract
+class LineGlyph(Glyph):
+    ''' Glyphs with line properties
+    '''
```
Comment (mattpap) [inline bokeh/models/glyph.py]: When introducing such classes, it would be reasonable to move related properties from concrete glyph models to here, so that we would have:
```py
@abstract
class LineGlyph(Glyph):
    ''' Glyphs with line properties
    '''

    fill_props = Include(ScalarFillProps, use_prefix=False, help="""
    The %s values for the horizontal directed area.
    """)
```
However, then we would have to provide a mechanism to fill in glyph names (like `area` in this case, as I copied this from `HArea` glyph), so it may be more trouble than it's worth at this point. Though if we get to statically typing this code with mypy, then we may be further encouraged to do so by the type-checker.

Comment (jeremy9959) [reply bokeh/models/glyph.py]: I considered this, and it does seem natural, but I was nervous about messing around too much with the properties system that I don't fully understand.  As it is,  it is inelegant that I had to match the mixin classes with the properties "by hand."

**Comment by mattpap** [comment]
@jeremy9959, can you add the code from the initial comment as an example under `examples/plotting/file`. This way we will be able to visually test this feature in future.

File: bokeh/models/glyphs.py [hunk 3]
```diff
@@ -124,7 +132,7 @@
 # General API
 #-----------------------------------------------------------------------------
 
-class AnnularWedge(XYGlyph):
+class AnnularWedge(LineGlyph, FillGlyph):
```
Comment (mattpap) [inline bokeh/models/glyphs.py]: Removing `XYGlyph` is incorrect, because you are removing vital properties and inheritance chain. Those `{Line,Fill,Hatch}Glyph` classes should be treated as mixins, so in the end the signature should look like this:
```py
class AnnularWedge(XYGlyph, LineGlyph, FillGlyph):
```
This is fine, because Python supports multiple inheritance. This obviously applies to all other changes.

**Comment by mattpap** [comment]
This failure is interesting and needs to be debugged:
```py
@pytest.mark.parametrize('glyph', [Line, Patch])
    def test_check_cdsview_filters_with_connected_error(self, glyph) -> None:
        renderer = bmr.GlyphRenderer()
        renderer.glyph = glyph()
    
        check = renderer._check_cdsview_filters_with_connected()
        assert check == []
    
        renderer.view.filters = [IndexFilter()]
        check = renderer._check_cdsview_filters_with_connected()
>       assert check != []
E       assert [] != []
```

File: bokeh/models/glyphs.py (hunk 3 above)
Comment (jeremy9959) [reply bokeh/models/glyphs.py]: OK, that makes sense - thanks.

**Comment by jeremy9959** [comment]
I'll also look into the error you flagged above.

**Comment by mattpap** [comment]
> I'll also look into the error you flagged above.

I would start with my earlier [comment](https://github.com/bokeh/bokeh/pull/9755#issuecomment-605633606), that may be the reason this failure. There are some other failures, e.g. in bokehjs' tests, that are definitively connect to this.

**Comment by jeremy9959** [comment]
On reflection I don't really know why I felt the need to kill XYGlyph, it doesn't surprise me that I broke things.

**Comment by jeremy9959** [comment]
I think I've reduced the problem to the lack of a baseline file -- the relevant CI error is:
```
tests/baselines/examples/plotting/file/theme_glyphs baseline doesn't exist
```

This is the same point I've reached in my other PR #9729.  So I need some help/guidance
on how to move forward.

**Comment by mattpap** [comment]
> `tests/baselines/examples/plotting/file/theme_glyphs baseline doesn't exist`

There should have been the contents of the baseline along that error message, so that you could copy it and commit together with the new example, but it's missing for some reason. I will need to investigate this.

**Comment by mattpap** [comment]
@jeremy9959, I fixed the PR.

**Comment by mattpap** [comment]
Test failure is unrelated.

**Comment by jeremy9959** [comment]
Does this need an update to the documentation?

**Comment by mattpap** [comment]
> Does this need an update to the documentation?

That would be nice, though I see there's none currently, so you would have to start from scratch.

**Comment by jeremy9959** [comment]
Well I seem to be home a lot these days so maybe I'll take that on.

**Comment by mattpap** [comment]
I'm going ahead and merge this. Please create a fresh PR if you decide to work on the docs.
